File: AppCore/Sources/AppCore/Selection/LockScreenSummary.swift
import Foundation

/// 잠금화면 표시용 요약 DTO
/// - lockscreen.md A 준수
public struct LockScreenSummary: Equatable, Sendable {
    /// 표시할 태스크 목록 (Top N)
    public let displayList: [DisplayTask]

    /// 카운터들
    public let counters: Counters

    /// 폴백 사유 (디버그/로깅용)
    public let fallbackReason: FallbackReason?

    public init(
        displayList: [DisplayTask],
        counters: Counters,
        fallbackReason: FallbackReason? = nil
    ) {
        self.displayList = displayList
        self.counters = counters
        self.fallbackReason = fallbackReason
    }
}

// MARK: - DisplayTask

/// 표시용 태스크 (프라이버시 적용됨)
public struct DisplayTask: Identifiable, Equatable, Sendable {
    public let id: String
    public let displayTitle: String
    public let isOverdue: Bool
    public let isDueSoon: Bool
    public let isRecurring: Bool
    /// 작업 중 여부 (G1 그룹)
    public let isDoing: Bool
    /// 우선순위 (P1~P4)
    public let priority: Priority

    public init(
        id: String,
        displayTitle: String,
        isOverdue: Bool = false,
        isDueSoon: Bool = false,
        isRecurring: Bool = false,
        isDoing: Bool = false,
        priority: Priority = .defaultPriority
    ) {
        self.id = id
        self.displayTitle = displayTitle
        self.isOverdue = isOverdue
        self.isDueSoon = isDueSoon
        self.isRecurring = isRecurring
        self.isDoing = isDoing
        self.priority = priority
    }

    /// P1 우선순위인지 여부 (편의 플래그)
    public var isP1: Bool {
        priority == .p1
    }
}

// MARK: - Counters

/// 카운터 집합
/// - lockscreen.md B 준수
/// - 필수: outstandingTotal, overdueCount, dueSoonCount, recurringDone/Total
/// - 선택: p1Count, doingCount, blockedCount
public struct Counters: Equatable, Sendable {
    /// 전체 미완료 수
    public let outstandingTotal: Int

    /// 지연(overdue) 수
    public let overdueCount: Int

    /// 임박(dueSoon) 수
    public let dueSoonCount: Int

    /// 반복 중 오늘 완료 수
    public let recurringDone: Int

    /// 전체 반복 수
    public let recurringTotal: Int

    /// P1 우선순위 태스크 수 (선택)
    public let p1Count: Int

    /// 작업 중(doing) 태스크 수 (선택)
    public let doingCount: Int

    /// 차단된(blocked) 태스크 수 (선택)
    public let blockedCount: Int

    public init(
        outstandingTotal: Int = 0,
        overdueCount: Int = 0,
        dueSoonCount: Int = 0,
        recurringDone: Int = 0,
        recurringTotal: Int = 0,
        p1Count: Int = 0,
        doingCount: Int = 0,
        blockedCount: Int = 0
    ) {
        self.outstandingTotal = outstandingTotal
        self.overdueCount = overdueCount
        self.dueSoonCount = dueSoonCount
        self.recurringDone = recurringDone
        self.recurringTotal = recurringTotal
        self.p1Count = p1Count
        self.doingCount = doingCount
        self.blockedCount = blockedCount
    }

    /// 빈 카운터
    public static let zero = Counters()
}

// MARK: - FallbackReason

/// 폴백 사유
public enum FallbackReason: String, Sendable {
    case noPinnedProject = "pinned_project_not_set"
    case pinnedProjectArchived = "pinned_project_archived"
    case pinnedProjectCompleted = "pinned_project_completed"
    case noTasks = "no_tasks"
    case allCompleted = "all_completed"
}

// MARK: - Empty State

extension LockScreenSummary {
    /// 빈 상태
    public static let empty = LockScreenSummary(
        displayList: [],
        counters: .zero,
        fallbackReason: .noTasks
    )
}

File: Android/LivePlan-Android/core/src/main/kotlin/com/liveplan/core/selection/LockScreenSummary.kt
package com.liveplan.core.selection

import com.liveplan.core.model.Task

/**
 * Lock screen display summary
 * Aligned with iOS AppCore LockScreenSummary
 *
 * Output of OutstandingComputer.compute()
 */
data class LockScreenSummary(
    /**
     * Top N tasks to display (max 3 for widget)
     */
    val displayList: List<DisplayTask>,

    /**
     * Counters for summary display
     */
    val counters: Counters,

    /**
     * Metadata about the computation
     */
    val metadata: Metadata = Metadata()
) {
    /**
     * Task with display information
     */
    data class DisplayTask(
        val task: Task,
        val maskedTitle: String,
        val group: PriorityGroup
    )

    /**
     * Priority group for selection
     * Order: G1 > G2 > G3 > G4 > G5 > G6
     */
    enum class PriorityGroup {
        G1_DOING,       // workflowState = DOING
        G2_OVERDUE,     // dueAt < now
        G3_DUE_SOON,    // 0 < dueAt - now <= 24h
        G4_P1,          // priority = P1
        G5_HABIT_TODAY, // habitReset recurring, today incomplete
        G6_OTHER        // remaining
    }

    /**
     * Summary counters
     */
    data class Counters(
        val outstandingTotal: Int = 0,
        val overdueCount: Int = 0,
        val dueSoonCount: Int = 0,
        val recurringDone: Int = 0,
        val recurringTotal: Int = 0,
        val p1Count: Int = 0,
        val doingCount: Int = 0,
        val blockedCount: Int = 0
    )

    /**
     * Computation metadata
     */
    data class Metadata(
        val dateKey: String = "",
        val scope: Scope = Scope.TODAY_OVERVIEW,
        val fallbackReason: FallbackReason? = null
    )

    /**
     * Selection scope
     */
    enum class Scope {
        PINNED_PROJECT,
        TODAY_OVERVIEW
    }

    /**
     * Fallback reason when pinned scope not available
     */
    enum class FallbackReason {
        NO_PINNED_PROJECT,
        PINNED_NOT_ACTIVE
    }

    companion object {
        /**
         * Empty summary for fail-safe fallback
         */
        val EMPTY = LockScreenSummary(
            displayList = emptyList(),
            counters = Counters()
        )
    }
}
